geoutils/georaster/spatial_tools.py

```diff
@@ -0,0 +1,466 @@
+    optimizing memory usage.
+
+    :param rasters: A list of geoutils Raster objects to be stacked.
+    :param reference: The reference index, in case the reference is to be stacked, or a separate Raster object \
+ in case the reference should not be stacked. Defaults to the first raster in the list.
+    :param resampling_method: The resampling method for the raster reprojections.
+    :param use_ref_bounds: If True, will use reference bounds, otherwise will use maximum bounds of all rasters.
+    :param diff: If True, will return the difference to the reference, rather than the DEMs.
+    :param progress: If True, will display a progress bar. Default is True.
+
+    :returns: The stacked raster with the same parameters (optionally bounds) as the reference.
+    """
+    # Check resampling method
+    if isinstance(resampling_method, str):
+        resampling_method = resampling_method_from_str(resampling_method)
+
+    # Select reference raster
+    if isinstance(reference, int):
+        reference_raster = rasters[reference]
+    elif isinstance(reference, gu.Raster):
+        reference_raster = reference
+    else:
+        raise ValueError("reference should be either an integer or geoutils.Raster object")
+
+    # Set output bounds
+    if use_ref_bounds:
+        dst_bounds = reference_raster.bounds
+    else:
+        dst_bounds = merge_bounding_boxes([raster.bounds for raster in rasters], resolution=reference_raster.res[0])
+
+    # Make a data list and add all of the reprojected rasters into it.
+    data: list[np.ndarray] = []
+
+    for raster in tqdm(rasters, disable=~progress):
+
+        # Check that data is loaded, otherwise temporarily load it
+        if not raster.is_loaded:
+            raster.load()
+            raster.is_loaded = False
+
+        # Reproject to reference grid
+        reprojected_raster = raster.reproject(
+            dst_bounds=dst_bounds,
+            dst_res=reference_raster.res,
+            dst_crs=reference_raster.crs,
+            dtype=reference_raster.data.dtype,
+            dst_nodata=reference_raster.nodata,
+        )
+
+        # Optionally calculate difference
+        if diff:
+            ddem = (reference_raster.data - reprojected_raster.data).squeeze()
+            ddem, _ = get_array_and_mask(ddem)
+            data.append(ddem)
+        else:
+            dem, _ = get_array_and_mask(reprojected_raster.data.squeeze())
+            data.append(dem)
+
+        # Remove unloaded rasters
+        if not raster.is_loaded:
+            raster._data = None
+
+    # Convert to numpy array
+    data = np.asarray(data)
+
+    # Save as gu.Raster
+    r = gu.georaster.Raster.from_array(
+        data=data,
+        transform=rio.transform.from_bounds(*dst_bounds, width=data[0].shape[1], height=data[0].shape[0]),
+        crs=reference_raster.crs,
+        nodata=reference_raster.nodata,
+    )
+
+    return r
+
+
+def merge_rasters(
+    rasters: list[gu.georaster.Raster],
```

Comment:
Same comments as for stack_rasters overall: make sure it works for any child class.